python/lsst/daf/butler/direct_query_driver/_driver.py

```diff
@@ -1222,3 +1224,33 @@ def _convert_dimension_record_results(
                     d["timespan"] = self._timespan_repr_cls.extract(m, name=timespan_qualified_name)
                 record_set.add(record_cls(**d))
         return DimensionRecordResultPage(spec=result_spec, next_key=next_key, rows=record_set)
+
+
+def _convert_data_coordinate_results(
+    spec: DataCoordinateResultSpec,
+    raw_rows: Iterable[sqlalchemy.Row],
+    next_key: PageKey | None,
+) -> DataCoordinateResultPage:
+    """Convert a raw SQL result iterable into a page of `DataCoordinate`
+    query results.
+
+    Parameters
+    ----------
+    spec : `DataCoordinateResultSpec`
+        Specification for the output values.
+    raw_rows : `~collections.abc.Iterable` [ `sqlalchemy.Row` ]
+        Iterable of SQLAlchemy rows, with `Postprocessing` filters already
+        applied.
+    next_key : `PageKey` or `None`
+        Key for the next page to add into the returned page object.
+
+    Returns
+    -------
+    result_page : `DataCoordinateResultPage`
+        Page object that holds a `DataCoordinate` container.
+    """
+    rows = [
+        DataCoordinate.standardize(cast(Mapping, row._mapping), dimensions=spec.dimensions)
```

Comment:
Yeah I don't think conceptually it's too terrible, since the `ResultSpec` we have available here owns the list of dimensions (`spec.dimensions`) and also owns the process of deciding what columns will be in the result (`spec.get_result_columns`).

Maybe I can add a method to ColumnSet (or ResultSpec-delegating-to-ColumnSet) that is equivalent to the `row[:len(spec.dimensions)]` bit of your code, and put it right next to `__iter__` so it's more local.  Or something similar to that -- basically explicitly keep track of which columns correspond to what somewhere.

Reply:
How do you feel about this?: https://github.com/lsst/daf_butler/pull/1014/commits/fed3cd97cf9dbd93c3a68acf7ce7136552302602